### PR TITLE
feat: introduce queue ✨ 

### LINKED ISSUE #463: No indication of the buy order has been rejected on Binance
## Version

v0.0.88 (unspecified)

## Description

Bot tried to put buy orders which has been rejected from Binance without the notice. User just see symbols with the trigger price clearly reached but no buy action is indicated.

## To Reproduce

It has happened while market was rapidly raising. Is it possible in time between bot measured the price for buy order and placed the buy order on the Binance the price just raised so much it was not possible to put this buy order with calculated price anymore?

Edit: see Additional context

## Expected Behaviours

There should be warning about the bot tried to put buy order and the buy order has been rejected on the Binance.

## Screenshots

There are only the log records indicating the buy action but no rejections logged:

image

## Additional context

Example of log record:

`{"name":"binance-api","version":"0.0.88","hostname":"73b0809731b1","pid":45,"gitHash":"unspecified","server":"binance","jobName":"trailingTrade","correlationId":"e96b2362-794a-4315-8ace-953726201893","level":50,"err":{"message":"Stop price would trigger immediately.","name":"Error","stack":"Error: Stop price would trigger immediately.\n    at /srv/node_modules/binance-api-node/dist/http-client.js:100:17\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async execute (/srv/dist/server.js:1:38463)\n    at async /srv/dist/server.js:1:8847\n    at async errorHandlerWrapper (/srv/dist/server.js:1:87322)\n    at async P (/srv/dist/server.js:1:7295)","code":-2010},"errorCode":-2010,"debug":true,"msg":"⚠ Execution failed.","time":"2022-08-10T13:00:49.789Z","v":0}`

HBARBUSD was actually not raising at the time this issue occured:

image

### PR DESCRIPTION
**Pull request by habibalkhabbaz**
## Description

This PR introduces the queue feature

- Each symbol has its own queue
- Fix #463 
- Fix order update from WebSocket by comparing `orderId`

## Related Issue

#451 
#463 

## Motivation and Context

After moving to WebSocket we encounter some small issues regarding the trailingTrade execution, we start thinking about using a queue to improve the bot #448

### CODE REVIEW

**Comment by habibalkhabbaz** [comment]
@chrisleekr 
It would be nice if you can test the performance of this PR in your Rasberry Pi because I don't have one xD

@uhliksk 
I added a fix for #463, you can test and check if it's solved.

**Comment by uhliksk** [comment]
@habibalkhabbaz Nice job. I'll try. Thank you.

@chrisleekr Please merge this #464 before my #462 as I can use this queueing mechanism instead of my simple queueing to solve #398 and #457. Thank you.

**Comment by uhliksk** [comment]
@habibalkhabbaz Logging is working fine.

image

**Comment by habibalkhabbaz** [comment]
> @habibalkhabbaz Logging is working fine.
> 
> 
> 
> image
> 
> 

Perfect!

**Comment by uhliksk** [comment]
@habibalkhabbaz Is this because of racing the queue?

```
{"name":"binance-api","version":"0.0.88","hostname":"f545a1de3f6a","pid":46,"gitHash":"unspecified","server":"cronjob","job":"trailingTradeIndicator","uuid":"b8a528ad-2651-492e-bf4f-4b7c7870bb75","level":50,"tag":"job-timeout","msg":"Failed to run the job within 20000ms.","time":"2022-08-11T15:00:13.013Z","v":0}
```

**Comment by habibalkhabbaz** [comment]
> @habibalkhabbaz Is this because of racing the queue?
> 
> 
> 
> ```
> 
> {"name":"binance-api","version":"0.0.88","hostname":"f545a1de3f6a","pid":46,"gitHash":"unspecified","server":"cronjob","job":"trailingTradeIndicator","uuid":"b8a528ad-2651-492e-bf4f-4b7c7870bb75","level":50,"tag":"job-timeout","msg":"Failed to run the job within 20000ms.","time":"2022-08-11T15:00:13.013Z","v":0}
> 
> ```

Hello @uhliksk 
I see it's happening to me too.
Let me check.
I think the problem coming from locking the symbol.
I am testing without locking now. If it's fine then I can solve it.

**Comment by habibalkhabbaz** [comment]
@uhliksk 
Done. Pushed a fix. 
Let me know if it's solved.

After implementing the queue, we don't need to do the trick of locking symbol in `trailingTrade`.
The queue will do the job.

@chrisleekr let me know if you confirm me on this too.

**Comment by uhliksk** [comment]
@habibalkhabbaz I found it was actually redis-server consuming too many resources. Even after the fix applied I had to restart redis-server manually.

image

**Comment by habibalkhabbaz** [comment]
@uhliksk 
Yes, there is a way to improve this. We have to re-use Redis connections based on the documentation:

https://github.com/OptimalBits/bull/blob/master/PATTERNS.md#reusing-redis-connections

By doing this it will decrease the usage of Redis too much.

redis-usage

However, this will report the following to Node.js process

```
Possible EventEmitter memory leak detected. 11 error listeners added to [Commander]. Use emitter.setMaxListeners() to increase limit
```

This is not a real memory leak and we can ignore it using the following, which will make the listeners unlimited:

```
require('events').EventEmitter.defaultMaxListeners = 0;
```

**Important note**: using the above will make us sacrifice by the memory leak warnings because we may don't see any real memory leak warnings in future.

What do you think @uhliksk ?
I think @chrisleekr can help us in this too.

**Comment by uhliksk** [comment]
@habibalkhabbaz I didn't analyzed this yet but I see CPU usage of redis-server is raising again and it's consuming same %CPU even after binance-bot is restarted. To reduce %CPU back to almost zero I have to restart binance-redis itself. This looks like the issue is something persistent on redis-server. Isn't the database on redis-server just growing without removing the old records?

**Comment by uhliksk** [comment]
@habibalkhabbaz It is growing fast.

image

**Comment by habibalkhabbaz** [comment]
@uhliksk 
Ignore my last message.

Thanks for the info. It helps me to debug.
I found the issue. Going to push a fix.

**Comment by habibalkhabbaz** [comment]
@uhliksk 
Done.
Now the Redis CPU usage must be better and the keys will not increase horribly.

**Comment by uhliksk** [comment]
@habibalkhabbaz Great job! Thank you.

**Comment by habibalkhabbaz** [comment]
@uhliksk 
If everything goes well, I think this PR will be ready to be reviewed by @chrisleekr 💪

**Comment by uhliksk** [comment]
Hello @habibalkhabbaz, there is no issue since last commit on my side. I think this PR is ready to be reviewed by @chrisleekr.

**Comment by chrisleekr** [comment]
@habibalkhabbaz @uhliksk 

Hi guys, 

It is great. :)

I will review/test it this weekend.

**Comment by chrisleekr** [comment]
Guys,

I've added Bull Board to the stack. 
You can now access Bull Board `/queue`.

To be saving space, I limit it to keeping 100 completed tasks.
As the job is fast, you won't see much in the Active tab.

As this board does not have authentication, I think we can add in later on.

So far the changes look good. I will continue to review/test.

image

**Comment by habibalkhabbaz** [comment]
Nice update @chrisleekr 🎉

**Comment by chrisleekr** [comment]
@habibalkhabbaz 

Yeah, that would be a good addition. :+1: 
Although, I think we can add that later on, not with this PR. 
We can even save the result of the job to debug.

This queue implementation is pretty stable. I can see a lot of benefits. 
Let's give one or two days more and see whether we can find any bug.

File: app/binance/__tests__/tickers.test.js [hunk 1]
```diff
@@ -66,16 +72,14 @@ describe('tickers.js', () => {
         return mockWebsocketTickersClean;
       });
 
-      mockExecuteTrailingTrade = jest.fn().mockResolvedValue(true);
+      // mockExecuteTrailingTrade = jest.fn().mockResolvedValue(true);
```
Comment (chrisleekr) [inline app/binance/__tests__/tickers.test.js:75]: Delete the line as it's commented.

**Comment by chrisleekr** [comment]
I reviewed the code @habibalkhabbaz 
Looks good. I think we are good to merge. 
I will merge in tomorrow if there is no issue found.

**Comment by habibalkhabbaz** [comment]
Thank you @chrisleekr
It seems good, there is no issue in my side till now.

**Comment by chrisleekr** [comment]
@habibalkhabbaz @uhliksk 

It's merged. Thanks for your contributions guys.

I have a few things to fix before releasing a new version.
Once all is resolved, I will release it today.

**Pull request merged**